File: client/src/modules/assetmanager/schemas/entity/entity.schemas.ts
/**
 * Entity Schemas
 *
 * Zod validation schemas for Entity model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/entity_models.py
 * - Schema: /server/apps/assetmanager/schemas/entity_schemas/entity_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/entity_subrouters/entity_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

/**
 * Entity type options - matches backend EntityType enum
 * Backend: class EntityType(str, Enum)
 */
export const EntityTypeEnum = z.enum(['fund', 'company', 'individual']);

/**
 * Entity role options - matches backend EntityRole enum
 * Backend: class EntityRole(str, Enum)
 */
export const EntityRoleEnum = z.enum(['OWNER', 'ADMIN', 'EDITOR', 'VIEWER']);

// ==========================================
// Entity Schema (Full Representation)
// ==========================================

/**
 * Entity schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class Entity(BaseModel)
 */
export const EntitySchema = z.object({
  id: z.number(),
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters'),
  entity_type: EntityTypeEnum,
  parent_id: z.number().nullable(),
  current_valuation: z.number().nullable(),
  organization_id: z.number(),
  cash_balance: z.number().default(0),
  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new entity (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class CreateEntity(BaseModel)
 */
export const CreateEntitySchema = z.object({
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters'),
  entity_type: EntityTypeEnum,
  parent_id: z.number().nullable().optional(),
  current_valuation: z.number().nullable().optional(),
  organization_id: z.number(),
  cash_balance: z.number().default(0).optional(),
});

/**
 * Schema for updating an entity (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class UpdateEntity(BaseModel)
 */
export const UpdateEntitySchema = z.object({
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters').optional(),
  entity_type: EntityTypeEnum.optional(),
  parent_id: z.number().nullable().optional(),
  current_valuation: z.number().nullable().optional(),
  cash_balance: z.number().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type EntityType = z.infer<typeof EntityTypeEnum>;
export type EntityRole = z.infer<typeof EntityRoleEnum>;
export type Entity = z.infer<typeof EntitySchema>;
export type CreateEntity = z.infer<typeof CreateEntitySchema>;
export type UpdateEntity = z.infer<typeof UpdateEntitySchema>;

// ==========================================
// Label Helpers
// ==========================================

/** Human-readable labels for entity types */
export const ENTITY_TYPE_LABELS: Record<EntityType, string> = {
  fund: 'Fund',
  company: 'Company',
  individual: 'Individual',
};

/**
 * Get human-readable label for an entity type
 * @param type Entity type enum value
 * @returns Display label string
 */
export const getEntityTypeLabel = (type: string): string => {
  return ENTITY_TYPE_LABELS[type as EntityType] || type;
};

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single entity
 * Backend equivalent: class EntityResponse(BaseModel)
 */
export type EntityResponse = {
  success: boolean;
  data?: Entity;
  error?: string;
};

/**
 * Response containing multiple entities
 * Backend equivalent: class EntitiesResponse(BaseModel)
 */
export type EntitiesResponse = {
  success: boolean;
  data?: Entity[];
  error?: string;
};

File: server/apps/assetmanager/schemas/entity_schemas/entity_schemas.py
"""
Entity Schemas

Pydantic schemas for the Entity model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from enum import Enum

# ==========================================
# Enums
# ==========================================

class EntityType(str, Enum):
    """Entity type options"""
    FUND = "fund"
    COMPANY = "company"
    INDIVIDUAL = "individual"

class EntityRole(str, Enum):
    """Entity organization member role options"""
    OWNER = "OWNER"
    ADMIN = "ADMIN" 
    EDITOR = "EDITOR"
    VIEWER = "VIEWER"

# ==========================================
# Entity Schema (Full Representation)
# ==========================================

class Entity(BaseModel):
    """Entity schema - full representation"""
    id: int
    name: str = Field(min_length=1, max_length=255, description="Entity name")
    entity_type: EntityType = Field(description="Type of entity")
    parent_id: int | None = Field(None, description="Parent entity ID")
    current_valuation: float | None = Field(None, description="Current valuation")
    organization_id: int = Field(description="Organization ID that owns this entity")
    cash_balance: float = Field(default=0, description="Cash balance")
    created_at: datetime
    updated_at: datetime | None = None
    
    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class CreateEntity(BaseModel):
    """Schema for creating a new entity"""
    name: str = Field(min_length=1, max_length=255, description="Entity name")
    entity_type: EntityType = Field(description="Type of entity")
    parent_id: int | None = Field(None, description="Parent entity ID")
    current_valuation: float | None = Field(None, description="Current valuation")
    organization_id: int = Field(description="Organization ID that owns this entity")
    cash_balance: float = Field(default=0, description="Cash balance")

class UpdateEntity(BaseModel):
    """Schema for updating an entity"""
    name: str | None = Field(None, min_length=1, max_length=255)
    entity_type: EntityType | None = None
    parent_id: int | None = None
    current_valuation: float | None = None
    cash_balance: float | None = None

# ==========================================
# Response Types
# ==========================================

class EntityResponse(BaseModel):
    """Response containing a single entity"""
    success: bool
    data: Entity | None = None
    error: str | None = None

class EntitiesResponse(BaseModel):
    """Response containing multiple entities"""
    success: bool
    data: list[Entity] | None = None
    error: str | None = None